### PR TITLE
Update LIP 0047 with genesis block processing

### PR DESCRIPTION
**Pull request by IkerAlus**

### CODE REVIEW

File: proposals/lip-0047.md
Comment (AndreasKendziorra) [inline proposals/lip-0047.md:652]: The return value should be checked, and block execution cancelled if the the return value is not `True`.

File: proposals/lip-0047.md
Comment (AndreasKendziorra) [inline proposals/lip-0047.md:648]: should't we abort the block execution if this function fails?

File: proposals/lip-0047.md
Comment (AndreasKendziorra) [inline proposals/lip-0047.md:644]: `snapshotStore(0).validators` instead of `snapshotStore(0).addresses`?

File: proposals/lip-0047.md
Comment (AndreasKendziorra) [inline proposals/lip-0047.md:656]:
```suggestion
    addresses = snapshotStore(0).validators[i].address
```

Comment (AndreasKendziorra) [reply proposals/lip-0047.md:656]: the same for line 691

**Review by AndreasKendziorra — CHANGES_REQUESTED** [review]
I see that some links are still pointing to the research forum. Maybe this could be fixed within this PR as well.

File: proposals/lip-0047.md
Comment (ishantiw) [reply proposals/lip-0047.md:554]: `validators` could be part of `genesisPoAStoreSchema` and we can remove `validatorsSubstore` as mentioned by @AndreasKendziorra

**Review by ishantiw — APPROVED** [review]

File: proposals/lip-0047.md
Comment (janhack) [reply proposals/lip-0047.md:509]: Consider using a short variable name such as `genesisAsset` or `assetPoA` instead of `genesisBlockAssetObject` for better readability and compactness below.

File: proposals/lip-0047.md
Comment (janhack) [inline proposals/lip-0047.md:511]: Wouldn't it be easier here and also better encourage code re-use to write:
For each entry in `genesisBlockAssetObject.validatorSubstore` perform the verification and execution logic of the authority registration command with the same parameters as input and `address` being the address of the sender account. Only the last step in the execution, callling `registerValidatorKeys`, is not done here as API calls should only be done in the genesis state finalization.

File: proposals/lip-0047.md
Comment (janhack) [inline proposals/lip-0047.md:518]: Wouldn't it be easier here and also better encourage code re-use to reference the same validation logic already described above:
- For `genesisBlockAssetObject.snapshotSubstore.activeValidators` apply the same verifications as for `trs.params.newValidators` in the update authority command verification.
- For `genesisBlockAssetObject.snapshotSubstore.threshold` apply the same verifications as for `trs.params.threshold` in the update authority command verification.

File: proposals/lip-0047.md
Comment (janhack) [inline proposals/lip-0047.md:542]: Above you sate about the `roundEndHeight`:
> Its initial value is set after the execution of the genesis block.

I think this should still be true and we should not assume that the genesis height is `0`. 

I would also remove the comment about the initialization that you have in the section above as it is redundant there.

File: proposals/lip-0047.md
Comment (janhack) [reply proposals/lip-0047.md:656]: Note that we also always overwrite the value `addresses`, but we actually want to append it to a list, right?

File: proposals/lip-0047.md
Comment (janhack) [reply proposals/lip-0047.md:648]: Note that in this LIP we sometimes use `BFTThreshold` and sometimes `BFT_THRESHOLD`. As this is a variable and not a constant, I would suggest to switch to camelCase, e.g., `bftThreshold`.

File: proposals/lip-0047.md
Comment (janhack) [reply proposals/lip-0047.md:542]: Ah, now I see that this value is overwritten later. It is normally not good practice to initialize something in between with an incorrect value, even if it is not used. 

Maybe consider setting it directly to the final value or first to the genesis height and then doing
```
    chainProperties.roundEndHeight = chainProperties.roundEndHeight + len(snapshotStore(0).validators)
```
below.

File: proposals/lip-0047.md
Comment (janhack) [inline proposals/lip-0047.md:674]: I think also after the pseudocode below we should list all functions that are used from other modules there. Currently, `setGeneratorList` does not appear in the list. In general, I would find it easier to once define the notation `moduleName.fct` and mention that in this LIP you call functions from the BFT, Random and Validator module. Then it the notation like `validatorsModule.setGeneratorList` or `Validators.setGeneratorList` makes it directly clear that this is not a local function defined in this LIP but an API.

**Review by janhack — CHANGES_REQUESTED** [review]
The general approach towards the genesis logic looks good. My suggestion would be to rather reference existing logic from the LIP instead of copy pasting it there.

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:503]: yes, hopefully, it will be merged in the next days.

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:545]: changed to the second suggestion

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:651]: fixed variable names as jan suggested.
Regarding Andreas question, I think that is already addressed in line 644

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:557]: `validatorsSubstore` removed as suggested

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md]: Added a correct formulation for the deserialization of the assets as Andreas suggested
Also, shortened `genesisBlockAssetObject` to `assetPoA` as Jan suggested

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:512]: I added a clarification sentence in this direction. This way we have the point of re-use, the strictly correct specs for the genesis and I don't need to revert changes  that are actually correct.

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:520]: I think in this case it is a bit dangerous to refer to that logic since for the update authority command verification the entries in the validator substore already exist and we do not need to check for uniqueness etc.

File: proposals/lip-0047.md
Comment (AndreasKendziorra) [reply proposals/lip-0047.md:504]: ok. I'll suggest to keep this comment open then and wait until the other PR is merged. Do you agree?

File: proposals/lip-0047.md
Comment (AndreasKendziorra) [inline proposals/lip-0047.md:510]:
```suggestion
Let `genesisBlockAssetBytes` be the bytes included in the `data` property of a deserialized entry in the block assets for the PoA module and let `assetPoA` be the deserialization of `genesisBlockAssetBytes` according to the `genesisPoAStoreSchema` schema. Then: 
```

Comment (AndreasKendziorra) [reply proposals/lip-0047.md:510]: Just a suggestion as the sentence is very long.

File: proposals/lip-0047.md
Comment (AndreasKendziorra) [reply proposals/lip-0047.md]: can be resolved

File: proposals/lip-0047.md
Comment (AndreasKendziorra) [reply proposals/lip-0047.md]: I think like this it is still not very clear as line 648 is just a comparison statement. But it's not clear that something fails here if `registerValidatorKeys` returns `False`. I would propose something like:

``` 
for every validator element in genesisBlockAssetObject.validatorSubstore.validators:
    if registerValidatorKeys(validator.address, validator.proofOfPossession, validator.generatorKey, validator.blsKey) != True:
        discard the block 
```

Maybe this functions becomes easier to understand if the logic is put into a function that returns True or False, and in the introduction you saysthat the block is discarded if the function returns `False`.

File: proposals/lip-0047.md
Comment (AndreasKendziorra) [reply proposals/lip-0047.md]: hm. I see 4 different cases where `[setBFTParameters](https://github.com/LiskHQ/lips/blob/c308e37920b31104da3f4ebd3eaebfdd8590a12f/proposals/lip-introduce_bft_module.md#execution-12)`, and already for the first I don't see clearly this will never happen. In particular, I do not see how line 644 prevents this.

File: proposals/lip-0047.md
Comment (AndreasKendziorra) [reply proposals/lip-0047.md]: can be resolved

File: proposals/lip-0047.md
Comment (AndreasKendziorra) [inline proposals/lip-0047.md:732]: again: could we update the links from the research forum to the LIPs repo for all published LIPs? :)

**Review by AndreasKendziorra — CHANGES_REQUESTED** [review]

File: proposals/lip-0047.md
Comment (janhack) [reply proposals/lip-0047.md]: Can be resolved.

File: proposals/lip-0047.md
Comment (janhack) [reply proposals/lip-0047.md]: From my side the comment can be resolved.

File: proposals/lip-0047.md
Comment (janhack) [reply proposals/lip-0047.md]: can be resolved

File: proposals/lip-0047.md
Comment (janhack) [reply proposals/lip-0047.md:671]: Can be resolved.

File: proposals/lip-0047.md
Comment (janhack) [inline proposals/lip-0047.md:279]:
```suggestion
* The `trs.params.name` property has to be unique with respect to the set of already registered authority names in the blockchain state, i.e. the entry with store prefix equal to `STORE_PREFIX_NAME` and store key equal to `trs.params.name` (serialized as a utf-8 encoded string) does not exist in the store.
```

File: proposals/lip-0047.md
Comment (janhack) [inline proposals/lip-0047.md:544]:
```suggestion
* Create an entry in the chain properties substore as specified beow.
```

File: proposals/lip-0047.md
Comment (janhack) [inline proposals/lip-0047.md:520]: We do not check the number of entries of the array as for the authority update command.

Comment (janhack) [reply proposals/lip-0047.md:520]: Regarding the comment above.

>  I think in this case it is a bit dangerous to refer to that logic since for the update authority command verification the entries in the validator substore already exist and we do not need to check for uniqueness etc.

Also in the authority update command we check address uniqueness. I think all checks are the same or which one specifically is different?

**Review by janhack — CHANGES_REQUESTED** [review]
There are a few minor open points left.

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:279]: updated

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:510]: updated

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:520]: Yes, the checks are equivalent but the command one is against the existing state store and here it is against the content of another property in the assets data. That is why I would prefer not to just point out to those checks. I have added a similar explanatory sentence as I did for checks for the `assetPoA.validators` though.

Good point regarding the number of entries, added.

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:521]: I fixed it by adding `where ⌊⋅⌋ is the floor function.` in the same line it is mentioned

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:523]: maybe I am missing something but I added the pairwise distinct condition above in this set of checks.

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:544]: removed any location detail

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:637]: updated here and above

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:732]: true, I thought I did it already

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md]: I think it is already clear that if this function does not return true, the processing must stop. 

Anyway, I went for your first suggestion here

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md]: It does not prevent that if fails, what it says is that if `setBFTParameters` (for whatever the reason) fails, then the processing of this block halts and it is discarded.

It is up to the sidechain dev to set the right values for the BFT parameters so that this function does not fail.

Comment (AndreasKendziorra) [reply proposals/lip-0047.md]: ok, I see that it is logically correct. But I think it's hard to get the logic from reading it. One reason is that it's not obvious at all that a setter function can fail. The second reason is now the inconsistency between calling `setBFTParameters` and calling `registerValidatorKeys`. For the latter one, the outcome of the call is explicitly used in the pseudo code. For the former one not. And it had to be done for `registerValidatorKeys` because "failing" is not well defined for this function.

File: proposals/lip-0047.md
Comment (AndreasKendziorra) [reply proposals/lip-0047.md:522]: I meant the suggestion created by the "suggestion" feature :) That means, rewriting this line to:

...there is an entry `validator` in the `genesisBlockAssetObject.validatorSubstore.validators` array with `validator.address == activeValidator.address`.

File: proposals/lip-0047.md
Comment (AndreasKendziorra) [inline proposals/lip-0047.md:518]: I know that you added the statement about the equivalence based on a comment from Jan. But I think it is misleading as it is not to 100% equivalent as you pointed out. Maybe "similar", "almost identical" or "almost equivalent". Or you could just drop it.

**Review by AndreasKendziorra — CHANGES_REQUESTED** [review]
Just one minor detail left.

File: proposals/lip-0047.md
Comment (janhack) [inline proposals/lip-0047.md:549]: I would suggest to run `prettier` on the schema to improve the spacing.

File: proposals/lip-0047.md
Comment (janhack) [inline proposals/lip-0047.md:574]: For me the `address` would be the primary identifier and I would therefore put that one first. Additionally, I would consider enforcing lexicographic order by address. This is not really required for the protocol, but I think it makes manual verification and also the programmatic verification (distinct addresses) easier.

**Review by janhack — CHANGES_REQUESTED** [review]
Looks good. I just have two minor comments.

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:503]: link to LIP 0060 has been added

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:730]: done

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:549]: good point, done

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:519]: changed to "similar"

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:523]: Now it is all clear, changed as suggested, thanks.

P.D: linking old suggestions doesn't seem to work

File: proposals/lip-0047.md
Comment (IkerAlus) [reply proposals/lip-0047.md:574]: good point, I fixed the schema and added the extra condition.

File: proposals/lip-0047.md
Comment (janhack) [reply proposals/lip-0047.md]: Okay, can be resolved.

**Review by janhack — APPROVED** [review]
Looks all good!

**Review by AndreasKendziorra — APPROVED** [review]

**Pull request merged**
